src/clients/java/src/main/java/com/tigerbeetle/NativeClient.java

```diff
@@ -36,55 +102,29 @@ private static void assertArgs(final byte[] clusterID, final String addresses,
     }
 
     private NativeClient(final long contextHandle) {
-        this.contextHandle = new AtomicLong(contextHandle);
-        this.contextHandleReferences = new AtomicLong(0L);
+        try {
+            this.handle = new NativeHandle(contextHandle);
+            this.cleanable = cleaner.register(this, () -> handle.close());
```

Comment:
>The cleaning action could be a lambda but all too easily will capture the object reference, by referring to fields of the object being cleaned, preventing the object from becoming phantom reachable. Using a static nested class, as above, will avoid accidentally retaining the object reference.

https://docs.oracle.com/javase%2F9%2Fdocs%2Fapi%2F%2F/java/lang/ref/Cleaner.html

Isn't this _exactly_ the problem we have here? `() -> handle.close()` is `() -> this.handle.close()`, so it does close over the entire `NativeClient`, preventing it from becoming unreachable?

Reply:
You're right, I hoped that `() -> handle.close()` only captured the `handle` instance, not `this`, but I'd need a scoped variable for that.

Lambda captures are so subtle that I'm not sure if this shortcut is worthy, I'll use the verbose equivalent with a `Runnable` action ... 

```
private static class CleanerAction implements Runnable {
    private NativeHandle handle;

    pub void run() { 
        // ...
    }
}
```